\name{limmas}
\alias{limmas}
\title{Differential expression analysis based on the negative binomial distribution}
\usage{
DESeq(object, test = c("Wald", "LRT"), fitType = c("parametric", "local",
  "mean"), betaPrior, full = design(object), reduced, quiet = FALSE,
  minReplicatesForReplace = 7, modelMatrixType)
}
\arguments{
  \item{object}{a DESeqDataSet object, see the constructor
  functions \code{\link{DESeqDataSet}},
  \code{\link{DESeqDataSetFromMatrix}},
  \code{\link{DESeqDataSetFromHTSeqCount}}.}

  \item{test}{either "Wald" or "LRT", which will then use
  either Wald significance tests (defined by
  \code{\link{nbinomWaldTest}}), or the likelihood ratio
  test on the difference in deviance between a full and
  reduced model formula (defined by
  \code{\link{nbinomLRT}})}

  \item{fitType}{either "parametric", "local", or "mean"
  for the type of fitting of dispersions to the mean
  intensity. See \code{\link{estimateDispersions}} for
  description.}

  \item{betaPrior}{whether or not to put a zero-mean normal
  prior on the non-intercept coefficients (Tikhonov/ridge
  regularization) See \code{\link{nbinomWaldTest}} for
  description of the calculation of the beta prior. By
  default, the beta prior is used only for the Wald test,
  but can also be specified for the likelihood ratio test.}

  \item{full}{the full model formula, this should be the
  formula in \code{design(object)}, only used by the
  likelihood ratio test}

  \item{reduced}{a reduced formula to compare against, e.g.
  the full model with a term or terms of interest removed,
  only used by the likelihood ratio test}

  \item{quiet}{whether to print messages at each step}

  \item{minReplicatesForReplace}{the minimum number of
  replicates required in order to use
  \code{\link{replaceOutliers}} on a sample. If there are
  samples with so many replicates, the model will be refit
  after these replacing outliers, flagged by Cook's
  distance. Set to \code{Inf} in order to never replace
  outliers.}

  \item{modelMatrixType}{either "standard" or "expanded",
  which describe how the model matrix, X of the GLM formula
  is formed. "standard" is as created by
  \code{model.matrix} using the design formula. "expanded"
  includes an indicator variable for each level of factors
  in addition to an intercept. for more information see the
  Description of \code{\link{nbinomWaldTest}}. betaPrior
  must be set to TRUE in order for expanded model matrices
  to be fit.}
}
\value{
a \code{\link{DESeqDataSet}} object with results stored as
metadata columns. These results should accessed by calling
the \code{\link{results}} function. By default this will
return the log2 fold changes and p-values for the last
variable in the design formula.  See \code{\link{results}}
for how to access results for other variables.
}
\description{
This function performs a default analysis through the
steps: \enumerate{ \item estimation of size factors:
\code{\link{estimateSizeFactors}} \item estimation of
dispersion: \code{\link{estimateDispersions}} \item
negative binomial GLM fitting and Wald statistics:
\code{\link{nbinomWaldTest}} } For complete details on each
step, see the manual pages of the respective functions.
After the \code{DESeq} function returns a DESeqDataSet
object, results tables (log2 fold changes and p-values) can
be generated using the \code{\link{results}} function. See
the manual page for \code{\link{results}} for information
on independent filtering and p-value adjustment for
multiple test correction.
}
\details{
The differential expression analysis uses a generalized
linear model of the form:

\deqn{ K_{ij} \sim \textrm{NB}( \mu_{ij}, \alpha_i) }{ K_ij
~ NB(mu_ij, alpha_i) } \deqn{ \mu_{ij} = s_j q_{ij} }{
mu_ij = s_j * q_ij } \deqn{ \log_2(q_{ij}) = x_{j.} \beta_i
}{ log2(q_ij) = x_j. * beta_i }

where counts \eqn{K_{ij}}{K_ij} for gene i, sample j are
modeled using a negative binomial distribution with fitted
mean \eqn{\mu_{ij}}{mu_ij} and a gene-specific dispersion
parameter \eqn{\alpha_i}{alpha_i}. The fitted mean is
composed of a sample-specific size factor \eqn{s_j}{s_j}
and a parameter \eqn{q_{ij}}{q_ij} proportional to the
expected true concentration of fragments for sample j. The
coefficients \eqn{\beta_i}{beta_i} give the log2 fold
changes for gene i for each column of the model matrix
\eqn{X}{X}. The sample-specific size factors can be
replaced by gene-specific normalization factors for each
sample using \code{\link{normalizationFactors}}.  For
details on the fitting of the log2 fold changes and
calculation of p-values see \code{\link{nbinomWaldTest}}
(or \code{\link{nbinomLRT}} if using \code{test="LRT"}).

Experiments without replicates do not allow for estimation
of the dispersion of counts around the expected value for
each group, which is critical for differential expression
analysis. If an experimental design is supplied which does
not contain the necessary degrees of freedom for
differential analysis, \code{DESeq} will provide a message
to the user and follow the strategy outlined in Anders and
Huber (2010) under the section 'Working without
replicates', wherein all the samples are considered as
replicates of a single group for the estimation of
dispersion. As noted in the reference above: "Some
overestimation of the variance may be expected, which will
make that approach conservative." Furthermore, "while one
may not want to draw strong conclusions from such an
analysis, it may still be useful for exploration and
hypothesis generation."

The argument \code{minReplicatesForReplace} is used to
decide which samples are eligible for automatic replacement
in the case of extreme Cook's distance. By default,
\code{DESeq} will replace outliers if the Cook's distance
is large for a sample which has 7 or more replicates
(including itself). This replacement is performed by the
\code{\link{replaceOutliers}} function. This default
behavior helps to prevent filtering genes based on Cook's
distance when there are many degrees of freedom. See
\code{\link{results}} for more information about filtering
using Cook's distance, and the 'Dealing with outliers'
section of the vignette. Unlike the behavior of
\code{\link{replaceOutliers}}, here original counts are
kept in the matrix returned by \code{\link{counts}},
original Cook's distances are kept in
\code{assays(dds)[["cooks"]]}, and the replacement counts
used for fitting are kept in
\code{assays(object)[["replaceCounts"]]}.

Note that if a log2 fold change prior is used
(betaPrior=TRUE) then expanded model matrices will be used
in fitting. These are described in
\code{\link{nbinomWaldTest}} and in the vignette. The
\code{contrast} argument of \code{\link{results}} should be
used for generating results tables.
}
\examples{
dds <- makeExampleDESeqDataSet(betaSD=1, n=100)
dds <- DESeq(dds)
res <- results(dds)
ddsLRT <- DESeq(dds, test="LRT", reduced= ~ 1)
resLRT <- results(ddsLRT)
}
\author{
Michael Love
}
\references{
DESeq2 reference:

Michael I Love, Wolfgang Huber, Simon Anders: Moderated
estimation of fold change and dispersion for RNA-Seq data
with DESeq2. bioRxiv preprint (2014)
\url{http://dx.doi.org/10.1101/002832}

DESeq reference:

Simon Anders, Wolfgang Huber: Differential expression
analysis for sequence count data. Genome Biology 11 (2010)
R106, \url{http://dx.doi.org/10.1186/gb-2010-11-10-r106}
}
\seealso{
\code{\link{nbinomWaldTest}}, \code{\link{nbinomLRT}}
}
